### PR TITLE
Many new functions for the Profiler JavaFX UI

### PR DESCRIPTION
**Pull request by PhRX**
Hi,

I've tried to stage all my changes in such a way that the consecutive commits are as clear as possible.
Some steps (such as the early reverts) are artifacts of my being new to git, so bear with me...

The commit "Refactor Step 8" though is the big one, which actually introduces all the new features.
The full list of UI changes can be found in that step, but for easy reference here it is again.

More features are coming in the near future, such as tree diffs, aggregation across threads (i.e. pull all traces from threads doing similar things together in a single tree), progress bar when loading a larger profile, copy/paste supprt, Tree View filtering, ...

List of functional changes :

- Allow opening multiple profiles
- Instead of Landing View, open log files or machines from Menu Bar
- Use Choice Box for selecting views
- Specify filters using UI instead of a string description
- Extended filter capabilities :
  * Also filter on Method Name
  * Added several comparison operators
  * Allow hiding threads which contain only "error" frames (works only in Tree View)
- Changing filter or removing filter will reapply filters to "original" profile which was filtered
- Collapse All / Expand All buttons added in Tree View
- Right-click on node in tree view for more expand/collapse options on the node itself
- Allow creating a diff of the flat views between two profiles (keyed by method name)
  * Diff view shows base and new numbers, and highlighted difference
- Export flat view or diff view as currently shown (i.e. filters taken into account) to a CSV file (using export button)
- Export (part of) a stack in tree view to a text file by right-clicking the node
- Show Thread Names in Tree View when known
- Show sample counts as well as percentages

### CODE REVIEW

**Comment by ikavalio** [comment]
Hi @PhRX, great pull request. 

I'm worrying about https://github.com/RichardWarburton/honest-profiler/pull/173/commits/2de095ae0506f3400b9fa8c9a4df5c23dc4c3b42. I don't think it's a correct way to solve the issue for Linux. The issue is caused by this lines:

```
jvmtiEvent events[] = {JVMTI_EVENT_CLASS_LOAD, JVMTI_EVENT_CLASS_PREPARE,
               JVMTI_EVENT_VM_DEATH, JVMTI_EVENT_VM_INIT, JVMTI_EVENT_COMPILED_METHOD_LOAD
 #ifdef GETENV_NEW_THREAD_ASYNC_UNSAFE
        ,JVMTI_EVENT_THREAD_START,
         JVMTI_EVENT_THREAD_END,
        JVMTI_EVENT_NATIVE_METHOD_BIND
 #endif
};
```

I haven't tested it, but I think placing ``` #ifdef GETENV_NEW_THREAD_ASYNC_UNSAFE``` before ```JVMTI_EVENT_NATIVE_METHOD_BIND``` will solve the problem and will not introduce any new one.

**Comment by PhRX** [comment]
Hi @ikavalio,

All I can reasonably say is that the fix (proposed by a colleague) works for me and my colleagues, but AFAIK we're all running Linux Mint, or maybe Ubuntu in one case.

I guess that if JVMTI_EVENT_THREAD_START and JVMTI_EVENT_THREAD_END are safe events on any platform, then your proposal makes more sense.

At any rate I'm fine with that commit being excluded, it's fairly trivial for me to keep it locally. Alternatively, feel free to add a commit with your proposed change to the pull request, or let me know if you want me to do that.

Apologies for the formatting issues BTW, which make the diff unclear. This was my first in-depth encounter with git so I suffered from some teething problems (CRLF in this case). I'm more or less up-to-speed now.

**Comment by ikavalio** [comment]
Hi @PhRX,

I've pointed out to this change only because it breaks existing functionality as it seems to me. Sorry I can't compile and test agent at the moment, but it looks like in your case controller won't be started at all under Linux. Controller will be started only if `GETENV_NEW_THREAD_ASYNC_UNSAFE` is undefined (for all platforms except FreeBSD and OS X by author's design). But after that commit it will defined for Linux as well:

```
#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__linux)
   #define GETENV_NEW_THREAD_ASYNC_UNSAFE
#endif
...

#ifndef GETENV_NEW_THREAD_ASYNC_UNSAFE
     if (CONFIGURATION->host != NULL && CONFIGURATION->port != NULL) {
         controller->start();
     }
 #endif
```

Could you please clarify that you can control your agent by TCP socket?

I do believe that it is safe to define JVMTI_EVENT_THREAD_START and JVMTI_EVENT_THREAD_END events for all platforms, at least I can't imagine any situation why this events should be disabled for Linux. I'm not original author of this code, but it looks like a wrong place for `ifdef`.

**Comment by PhRX** [comment]
Hi @ikavalio,

I fixed it as per your suggestion. All tests pass, and the start=1 autostart feature still works on Linux as well.

Thanks for the clarifications !

**Comment by nitsanw** [comment]
I've started looking through the PR, it is very large and the work is appreciated.
It would make it easier to review this work if better discipline was exercised around style/white-space/formatting changes which increase the noise. This is mostly a fault of the review tools IMO, but still makes the reviewer life harder.

File: src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java [hunk 1]
```diff
@@ -163,19 +167,23 @@ private void emitProfile()
...
-                final ThreadMeta meta = metaByThreadId.get(threadId);
-                final NodeCollector collector = node.getValue();
-                if (meta == null) {
-                    return new ProfileTree(threadId, collector.normalise(methodByMethodId::get), collector.getNumberOfVisits());
-                }
-                return new ProfileTree(meta, collector.normalise(methodByMethodId::get), collector.getNumberOfVisits());
-            })
-            .sorted(sortBySampleCount)
-            .collect(toList());
+        return treesByThreadId.entrySet().stream().map(node ->
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java [hunk 2]
```diff
@@ -110,9 +110,12 @@ private void collectStackFrame(boolean endOfTrace, StackFrame stackFrame)
         {
             // might be null if the method hasn't been seen before
             // if this is the case then the handle(Method) will patch it
-            currentTreeNode = treesByThreadId.compute(currentThread, (id, previous) -> {
+            currentTreeNode = treesByThreadId.compute(currentThread, (id, previous) ->
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java (hunk 1 above)
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/filters/ClassNameFilter.java [hunk 3]
```diff
@@ -1,77 +1,35 @@
 /**
  * Copyright (c) 2014 Richard Warburton (richard.warburton@gmail.com)
  * <p>
- * Permission is hereby granted, free of charge, to any person obtaining a
- * copy of this software and associated documentation files (the "Software"),
- * to deal in the Software without restriction, including without limitation
- * the rights to use, copy, modify, merge, publish, distribute, sublicense,
- * and/or sell copies of the Software, and to permit persons to whom the
- * Software is furnished to do so, subject to the following conditions:
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/filters/ClassNameFilter.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/filters/TimeShareFilter.java [hunk 4]
```diff
@@ -78,62 +122,24 @@ private void filterFlatProfile(Profile profile)
     @Override
     public boolean equals(Object o)
     {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (this == o)
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/filters/TimeShareFilter.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java [hunk 5]
```diff
@@ -40,10 +40,25 @@ public Method(long methodId, String fileName, String className, String methodNam
...
+        this.methodId = methodId;
+        this.fileName = fileName;
+        this.className = className;
+        this.methodName = methodName;
+    }
+
     private String formatClassName(String className)
     {
         if (className.isEmpty())
+        {
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java [hunk 6]
```diff
@@ -78,8 +96,14 @@ public String getMethodName()
     @Override
     public boolean equals(Object o)
     {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (this == o)
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/Method.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/StackFrame.java [hunk 7]
```diff
@@ -63,8 +62,14 @@ public long getMethodId()
     @Override
     public boolean equals(Object o)
     {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (this == o)
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/StackFrame.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java [hunk 8]
```diff
@@ -47,8 +47,14 @@ public String getThreadName()
     @Override
     public boolean equals(Object o)
     {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (this == o)
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java [hunk 9]
```diff
@@ -67,12 +73,24 @@ public void accept(LogEventListener listener)
...
+    }
+
     @Override
     public String toString()
     {
-        return "ThreadMeta{" +
-            "threadId=" + threadId +
-            ", threadName=" + threadName + 
-            '}';
+        return "ThreadMeta{threadId=" + threadId + ", threadName=" + threadName + '}';
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/parser/ThreadMeta.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java [hunk 10]
```diff
@@ -90,8 +110,14 @@ private boolean isLogFile(final Path path)
     @Override
     public boolean equals(Object other)
     {
-        if (this == other) return true;
-        if (other == null || getClass() != other.getClass()) return false;
+        if (this == other)
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java [hunk 11]
```diff
@@ -105,11 +131,23 @@ public int hashCode()
     @Override
     public String toString()
     {
-        return "VirtualMachine{" +
-            "userDir='" + userDir + '\'' +
-            ", agentLoaded=" + agentLoaded +
-            ", displayName='" + displayName + '\'' +
-            ", id='" + id + '\'' +
+        return "VirtualMachine{"
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/sources/VirtualMachine.java]: whitespace/style

File: src/main/java/com/insightfullogic/honest_profiler/ports/javafx/UserInterfaceConfigurationException.java [hunk 12]
```diff
@@ -1,32 +1,32 @@
 /**
  * Copyright (c) 2014 Richard Warburton (richard.warburton@gmail.com)
  * <p>
- * Permission is hereby granted, free of charge, to any person obtaining a
- * copy of this software and associated documentation files (the "Software"),
- * to deal in the Software without restriction, including without limitation
- * the rights to use, copy, modify, merge, publish, distribute, sublicense,
- * and/or sell copies of the Software, and to permit persons to whom the
- * Software is furnished to do so, subject to the following conditions:
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/ports/javafx/UserInterfaceConfigurationException.java]: whitespace/style

**Comment by PhRX** [comment]
Hi @nitsanw,

I fully understand your frustration, and I apologize for the formatting changes, which may seem arbitrary.

In my defense however, a number of points I'd like to make : 
- The formatting of the source material seems variable. The maximum length of the lines before wrapping is as far as I can tell not constant. This is undoubtedly a natural consequence of having various contributors with varying IDEs.
- I have checked in the settings for my Eclipse formatter. I did take quite some time to try and get the settings such that any reformatting would affect the source as little as possible - i.e. I tried to craft it so that the vast majority of the conventions on this project are respected. The places where it doesn't are either an oversight or mistake on my part, or due to the limitations of the formatting options in Eclipse.
- I have Eclipse automatically format the file when I save. This, in the long run, makes for the most stable history, since any formatting mistakes get automatically corrected to a standard. Also, since I want to proceed really really fast, and the formatting rules for this project does not conform to my current "style" (which is 90% determined by the local dev team here, and 10% of my own "quirks"), auto-formatting is the easiest way of ensuring the format is fairly stable.\
- I paid special attention to any code in the core part, since I try to change that as little as possible. The FX part, I cared less about, since the volume of my changes would make diffing a pretty hairy business by any standards. Also, the first batch of changes (called the refactoring in the commit comments) was made in a different VCS, because I wasn't sure at that point whether my changes would be worth contributing...

As I wrote to @RichardWarburton at some point, I think formatting/whitespace/... should be treated much more like skinning by all development tools : when you're coding and viewing, the code should be auto-formatted to your own standards. Whenever published/checked in, the code should be auto-formatted to the repository standard. But, of course, that ain't the current way of things :(

So again, apologies for the frustration. I'll turn the format-on-save off from this point onwards, to try and ensure more consistency with what is already there...

**Comment by PhRX** [comment]
@nitsanw Alternatively, if you have eclipse formatting settings (or for some IDE) which do conform to what you need, please feel free to send them my way, that'd be optimal AFAICS.

**Comment by nitsanw** [comment]
"I have Eclipse automatically format the file when I save. This, in the long run, makes for the most stable history, since any formatting mistakes get automatically corrected to a standard. Also, since I want to proceed really really fast, and the formatting rules for this project does not conform to my current "style" (which is 90% determined by the local dev team here, and 10% of my own "quirks"), auto-formatting is the easiest way of ensuring the format is fairly stable."
I think this is the culprit. I believe there is an option to reformat only code which you touch.
Re-formatting the licence blurb and existing toString methods is just making a large change larger.
If you do want to tackle formatting it is better done IMO as separate, explicitly non-functional PRs, but that's extra work so mostly people skip it.
I would appreciate it if you worked through the diff and reverted any formatting changes. You are in a better position to tell what's whitespace and what isn't, but I understand if you don't have time for that.

**Comment by nitsanw** [comment]
@PhRX to be clear, I don't care about formatting, just the overheads of reviewing. If tools hid all the semntically neutral choices from me I'd be happier.

**Comment by PhRX** [comment]
@nitsanw OK I changed the setting to only format edited lines.

I'm not very familiar with github (or git for that matter) yet so quick question : you just posted a number of diffs as reviews on this pull request. Are these "proposed changes" I can accept somehow on github, or do I have to manually check and repair the diffs in my IDE for each file ?

**Comment by nitsanw** [comment]
@PhRX Further note, I think it's great you're investing your effort here. I also understand you are tacking quite a few issues as you find them. Please consider that smaller PRs are MUCH MUCH easier to review and merge (on all project, not just this). So if you have bits of this PR (or future PRs) which can be separated to different PRs, it would help if you did that.

**Comment by nitsanw** [comment]
@PhRX these are comments on the diff, not changes. You can choose to act on those, or discuss them or ignore them. I add them to mark bits of the review I need not look at at this point.

**Comment by PhRX** [comment]
@nitsanw Not being familiar with github, pull requests etc., I basically staged the refactoring into separate commits which were as small and coherent as possible. This was after I had a couple of days of changes already in the local VCS. The idea was that each commit in that refactor could hopefully be reviewed separately and more easily.

However I didn't know yet that commits "pile up" in the PR, so it all becomes one big effort from review point anyway.

After the initial "big push" I've been committing as nicely as possible - i.e. one "real change" at a time, with as few files affected as possible. I still don't see however how I can translate this to smaller PRs, since the commits will be bunched up into it. Short of branching for each change of course, but then the VCS overhead on development starts to become prohibitive.

Not having reviewed anything on github, I don't have any idea what "works" and "doesn't work" as workflow. I'll google some articles on git(hub) workflow.

**Comment by PhRX** [comment]
@nitsanw OK having thought over what is the most efficient use of my time, I propose that I fix the formatting to be 100% consistent with the "original code" for all classes outside the ports.javafx package. That should make it easily reviewable.

In ports.javafx, I've pretty much rebuilt most of the UI anyway, reusing only a couple of the components. The amount of changed code is big enough to make reviewing by looking at diffs impractical IMHO. Just trying it out and seeing whether things work as expected would take a lot less time (keeping in mind that tree filtering is broken, if I have time left I'll fix that today).

**Comment by PhRX** [comment]
@nitsanw I think I reverted all unnecessary formatting differences from unchanged lines in the core package. Thanks for reviewing BTW :)

File: src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java [hunk 13]
```diff
@@ -21,27 +21,31 @@
  **/
 package com.insightfullogic.honest_profiler.core.collector;
 
+import static java.util.Comparator.comparing;
```
Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: As far as I can see all changes to this file are whitespace/style

Comment (nitsanw) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: Sorry, confuse by GHub.
Imports reformatting is still noise.

Comment (PhRX) [inline src/main/java/com/insightfullogic/honest_profiler/core/collector/LogCollector.java]: Sorry about that. The meaningful bit amounts to :

136c140,143
<         metaByThreadId.put(newThreadMeta.getThreadId(), newThreadMeta);
---
>         metaByThreadId.merge(
>             newThreadMeta.getThreadId(),
>             newThreadMeta,
>             (oldMeta, newMeta) -> oldMeta.update(newMeta));

**Comment by RichardWarburton** [comment]
Hi, just a head's up.

I'm happy to review this PR at some point. But with 8000 lines total diff I don't see time for it in my schedule soon. That's not to say that the work isn't appreciated, but for future reference the best way (for both you and me) to get changes into this project are a series of small, single-feature changes. Things that could be reviewed quickly and easily.

**Comment by PhRX** [comment]
Hi @RichardWarburton,

Yes, now that I'm a bit more familiar with the github workflow I see how I can do this both technically and organizationally, and I'm committed to doing just that, get the future changes in nicely bite-sized (I've been doing that at commit/push level for a while now). The most important realization for me is that a PR actually keeps growing with new commits (I originally thought it grouped the commits at the PR point and that only explicit additions to the PR would be added, i.e. some sort of lightweight fork I guess).

Unfortunately, for this particular PR that boat has long sailed, and I'm painfully aware I should have gone about it a bit differently...

To resolve I'm working towards a particular milestone which I have in mind - which is the implementation of the Tree Diff view, and some refactoring to make the Controllers a lot cleaner.
I'll mark that point with a label, and fork (I guess, but I'll do some research into what is the best "git-way-of-doing-things" in case of frequent changes).

After that, the reviewing effort should be simplified somewhat : 
* The code in the core package can be reviewed using the classical diff-based reviewing. I've taken pains to ensure there's no "noise" left from reformatting or styling differences, and to keep it 100% backward compatible (i.e. if you were to move the core code into the code before the fork, it should just work).
* The code in the javafx package should probably be reviewed at that point "on its own merit", i.e. diffs will probably not help at all - correctness would possibly be best accounted for by just trying it out, and style/code quality can be determined by looking at the classes on themselves, which should be reasonably fast since I'll refactor for clarity and simplicity at some point before the milestone.

In the meantime I hope some people at least get some joy from the new features...

Thanks, and best wishes,
Philip

**Comment by nitsanw** [comment]
@PhRX please let me know when you are done and I'll review

**Comment by PhRX** [comment]
@nitsanw Much appreciated ! If all goes well (but does it ever...) II hope to finish the "milestone" tomorrow.

I've already cleaned up/refactored a *lot*, and the code has become simpler and controllers less cluttered. The last major bit left over is cleaning up (and slightly fixing) the renditions of the columns in the tables.
But that's for tomorrow :)

**Comment by PhRX** [comment]
Hi @nitsanw and others,

I'm done for the milestone I set myself, which means I won't add anything to this PR anymore unless by explicit request from reviewers.

I refactored to eliminate duplicate code as much as possible, documented the abstract controller superclasses, and tried to homogenize the naming as much as possible.

As a result, the javafx code should be a lot more readable, and concise.

There's one outstanding issue I'm aware of: the context menus on the comparison (or diff) tree view are having issues and are only appearing sporadically. There's an underlying bug somewhere I'm sure but I wasn't able to locate it. I have to leave that for next monday...

Xeers,
Philip

File: build.gradle [hunk 14]
```diff
@@ -0,0 +1,92 @@
+plugins {
```
Comment (RichardWarburton) [inline build.gradle]: Please don't just change the build system to gradle. This kind of change has basically no value and is impossible to review when combined with other codebase changes.

File: Eclipse_Neon_4.6.2_PHRX_Formatter.xml [hunk 15]
```diff
@@ -0,0 +1,313 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (RichardWarburton) [inline Eclipse_Neon_4.6.2_PHRX_Formatter.xml]: We should define a common project style, but not in an eclipse-specific format. Something like checkstyle would be used. My proposal would be to leave the style changes to another PR.

**Comment by PhRX** [comment]
Hi,

Re: Gradle : does providing the extra alternative break anything? My aim was not to change it, but to have it as an alternative for anyone who has their tooling set up for gradle rather than maven. My own first checkin was very neutral. The second commit contained changes made by a colleague, I think it didn't add anything that would break the maven build ?
Also, the gradle wrapper mechanism (./gradlew run) makes it very easy for anybody without any tooling to run the project straight away (well the java part at least).

Re: Formatter: I'm ok with defining a common style using eg checkstyle. However I don't think checkstyle contains any actuall reformatting tools, only checkers ? The idea of checking in my formatter (which can be updated very easily) is to help people who happen to be using eclipse to maintain the format. I'm willing to construct equivalent settings for IntelliJ or any other IDEs and check those in too.

Generally I try to follow an inclusive approach - which is to provide tools or configuration for all kinds of setups rather than singling out any particular one as the "One Ring To Rule Them All" :) . I know it's a bit more work and might be the occasional source of an issue but the payoff should to my mind be that more people might pick it up more quickly and contribute. I know that if proper formating settings had been available to me when I started out on this, the formatting issues with my early commits would never have occurred.

Anyway, It's ok for me if the gradle or eclipse files are not integrated. I guess I could maintain them pretty easily on a separate fork just for that purpose.

**Comment by PhRX** [comment]
I figured out the issue with the context menus last friday and fixed it (and i18ned the menu text in the process). What would be the way to go for these fixes: check them in on a new branch from my fork?  Or wait until the PR is closed ?

**Comment by RichardWarburton** [comment]
@PhRX 

Thanks for your reply, comments inline. If possible in future replying to specific comments directly rather than on the main thread might help, but I didn't want to just ignore your reply so I've replied to it here. I'll try to find some time this week to review a bit more of the PR, but I'm very busy atm so it might take some time.

re: build system - The reason I don't want to accept multiple build systems is that it makes the whole project much harder to maintain going forward. If someone has a pull request that touches the build system they have to test it with both and maintain both. if someone reports a compilation, testing or heck even a runtime issue I need to know what build system to reproduce it with. Not having a canonical build script that can build a project is a big fail IMHO. I really wouldn't be happy maintaining the project with multiple build systems.

I think the same thing with maintaining IDE-specific configuration. It would be much better to use an Independent tool that can be imported into any IDE and then maintain the configuration in one place, rather than in multiple places.

re: merging menus/i18n: I think I would appreciate not merging in any more changes until we can get a handle on the existing PR. Having said that if the i18n stuff helps cleanup any other code or helps you improve things more easily then I'm ok with it being merged into this PR.

**Comment by PhRX** [comment]
@RichardWarburton 

Thanks for your insights !

I'll maintain the gradle build stuff on a new fork then. Easy enough for me to manage in that case and it doesn't affect the main project at all. I'll always merge from the master to the fork so that other than the gradle stuff there's no differences. And any other work I do can be done on branches from that fork, and I can launch my PRs from there (and I should be able to exclude any gradle-related commits from the PR).

Should I delete the gradle files from the PR myself, or can you, when eventually merging the PR, just choose to leave them out ? I.e. not 100% sure whose responsibility this is.

I do hope that for maven you have something lined up along the lines of the https://github.com/shemnon/javafx-gradle javafx gradle plugin, which, although no longer maintained, still works and makes it easy to create e.g. an "all in" package (including a runtime java) which has linux and/or windows executables starting up the application. It's great for providing a no-config deployable for end users.

I also checked and found out that it's possible to generate an eclipse formatter from checkstyle configuration (see http://stackoverflow.com/questions/984778/how-to-generate-an-eclipse-formatter-configuration-from-a-checkstyle-configurati) so yes, having just a checkstyle config checked in makes more sense with that in mind. Same question though - do I delete the eclipse formatter config in the PR or can the PR merger just leave it aside ?

Finally - yes I'd like to integrate the context menu fix + I18N, since it does make the superclass code a bit more consistent (I introduced 2 slightly different ways of configuring columns trying to fix the issue, only the issue was something else entirely; the code is a little less consistent as a result). I need to clean it up, I'll push the changes in an hour or so unless I get feedback in the meantime not to do that after all...

**Comment by RichardWarburton** [comment]
Hi @PhRX, 

re: gradle I think it would be better if you removed it. Easier if it's just you editing the PR ;)

I don't know if there's a maven equivalent to the plugin you linked. I don't even object necessarily to switching the build to gradle as such. I just don't want to have two build systems or worry about build system differences in this PR. But I think such a move should probably be an issue that we discuss separately.

re: other points: awesome!

**Comment by PhRX** [comment]
To any potential reviewers : In the meantime I've reworked quite some things, and everything is now on the "integration" branch of my fork.

If nobody has started reviewing this PR yet, then it might make sense to merge my integration into this PR, since the code is more consistent still, and has a number of important improvements and fixes (see below).

I think it won't be (much) more work to review this new state - there is a bit more code, but everything has been streamlined a lot so the important bits of code tend to be concentrated, with little unnecessary duplication, and it should be easier to read.

Also, I documented (javadoc) a number of methods/classes already, which should make things a bit easier. I'll do more documentation in the near future, hopefully.

Please let me know what you think - merge into the PR or wait for the current PR to be accepted before submitting a new one ?

Here's a list of the most important changes : 

* FileLogSource now has a limited-size buffer which is a moving window over the file. Latest version also is fixed for the issues which arose from not taking into account that position could be set backwards in case of partial reads
* Top-level frames of stacks are now aggregated correctly - inside a thread there can be multiple top-level frames
* Time is now processed and aggregated, and shown. All views now have consistent data
* Aggregation happens at "fully qualified method name" (FQMN == FQCN + method name) level, so a method shows up only once in the flat view. Everything is in place to make it a user option to aggregate on method id instead, or FQMN + line nr, or file + bci or whatever.
* LogCollector / Profile / ... has been replaced by LeanLogCollector / LeanProfile / ... In this new setup, the collector does much less aggregating. An emitted profile is passed to a background task which turns it into an AggregationProfile (starting point for aggregations). The result is published to the front-end by the task. This effectively decouples the thread responsible for parsing / profile emission and the FX thread for updating the UI.
* The LeanLogCollector now operates on a pull-model : the front-end signals it'd like a new profile, and only then the collector will emit one. This means the emission rate can now easily be controlled from the front-end. The collector won't emit profiles when not asked for, which saves on processing for that thread.
* Filtering is now much more generic, and works on all columns - although there are some semantic issues left with tree filtering; and for time filtering the input still is in nanos whereas the display is in ms, so that's to be fixed.
* The Controllers have been refactored and share a lot more code at different levels, making them much easier to read and maintain. Column & field names have been standardized internally for easier maintenance as well.
* Monitoring or opening the same log now is blocked.

**Comment by PhRX** [comment]
In the meantime I've added new features on the integration branch on my fork (by branching then merging, to somewhat separate the various features and fixes). If nobody has started reviewing this PR, wouldn't it make sense to merge integration into it ? If anything, there's more documentation, and the code should mostly have become cleaner and clearer. On the minus side, yes, there's even more of it.

**Comment by PhRX** [comment]
Obsoleted by #184
